<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=iso-8859-1" />
<meta name="Last-modified" content="Mon, 10 Dec 2010 01:03:33 GMT"/>
<title>Amino Acid Conservation Service (AACon) command line executable help page</title>
<link href="ws.css" rel="stylesheet" type="text/css" media="screen,  projection, handheld, tv" />
<link rel="stylesheet" type="text/css" media="print" href="print.css"/>

<script type="text/javascript" src="prototype-1.6.0.3.js"></script>
</head>
<body>
<div id="page">
<div id="banner"><table> 
<tr><td style="width:130px;"><a href="http://www.dundee.ac.uk"><img class="logo" src="images/uod_lt.gif"  alt="University of Dundee"  title="University of Dundee" longdesc="http://www.dundee.ac.uk"/></a></td>
<td class="bg"><h1><span style="color:#1ACC1A" class="headeru">A</span>mino <span style="color:#80B3E6" class="headeru">A</span>cid <span style="color:#AE5439" class="headeru">Con</span>servation Service</h1></td>
</tr>
</table>
</div><!-- banner end-->
<div id="wrapper">
<div id="panel">
	<a href="index.html">Home</a>
<a class="selected" href="command.html" title="Click to view more information on the command line version of AACon" >Standalone</a>
	<a href="webservice.html" title="Click to view more information on AACon SOAP web service and how to use it" >Web Service</a>
	<a title="Click to view more information on AACon Java library" href="library.html">Library</a>	
	<a href="download.html" title="Click to download AACon" >Download</a>
	<a href="http://www.compbio.dundee.ac.uk">Barton Group</a></div>
<!-- panel end-->
<div id="content">
<h2 id="headtitle">AACon Standalone Executable  </h2>
<p>This page provides some tips on using AACon command line tool.</p>
<strong>Outline</strong>
<ul>
  <li><a href="#exec">Executing AACon</a></li>
  <li><a href="#in">The input</a></li>
  <li><a href="#out">The output </a></li>
  <li><a href="#run">Calculating conservation</a></li>
  <li><a href="#gap">Using custom gap character</a> </li>
  <li><a href="#cpulim">Limiting AACon to X number of CPU </a></li>
  <li><a href="#smerfs">Running SMERFS with custom parameters </a></li>
  <li><a href="#fileout">Recording the output to the file </a></li>
  <li><a href="#execdet">Accessing execution details </a></li>
  <li><a href="#normal">Results normalization</a> </li>
  <li><a href="#largeal">Conservation for large alignments (Java VM running out of memory) </a></li>
</ul>
<hr />
<p>&nbsp;</p>
<h3><a name="exec" id="exec"></a>Executing AACon</h3>
<p>Type java -jar &lt;AACon jar file name&gt;<span class="code">java -jar aacon.jar </span></p>
<p>This will print AACon help.  </p>
<h3><a name="in" id="in"></a>The input</h3>
<p>Input to AACon can be either </p>
<ul>
  <li>The list of aligned FASTA formatted sequences </li>
  <li>The Clustal formatted alignment</li>
  </ul>
<p>FASTA example</p>
<p class="code">&gt;UniRef90_Q5VFX<br />
MEWLKIALVAFSGRGNDVDYPCSILGRVANNDKELVNILRNGFFLLTYRMNFSPLPHSSVTSDKGWGCLVRS<br />
SQMLLAHALWRYSANDCRLDHFRDMDTEDSTPFSLHKMVRAVMKKADVFRPEYWTPSQGCEAIRCCVNNAVD<br />
RKLIPPIRVVVCSQGCLLAREICSNLEFGTVLILAPMRCGASRRMTQMMFFSLEHLLHSSACIGVVGGVPQR<br />
SYYILGTSGQRLLYLDPHCMTQEALVSSHAEKAGVVTVTASLVKSVRWDCVDTSCFLGFLVDSFAEWLELRT<br />
CLEELQRRGMEQLLCVDDGVAVGLVDEEIAGWPSEEDVAE<br />
&gt;UniRef90_Q4VBK1<br />
---------------PDTDEPVWILGACYNVKTKKSELLSDSRLWFTYRKKFSPIGGTGPSSDAGWGCMLRC<br />
GQMILAQALWRWDPEKHQPKEYQRFLDKKDSCYSIHQMAQMGVGE-GKSVGEWYGPNTVAQVLKKL----AL<br />
FDDWNSLSVYVSMDNTVVIEDIKKLCDWRPLLLVIPLRMGINS-INPVYIQALKECFKMPQSCGVLGGKPNL<br />
AYYFIGFIDDELIYLDPHTTQQA-VDTESGSAVDDQSFHCQRPHRMKITSLDPSVALGFFCKSEEDFDSWCD<br />
LVQ-------QELLKKRNLRMFELVEKHPSHWPPF-----<br />
&gt;UniRef90_A7RTH8<br />
---MDAACVTYEGVSHETEEDVWILGKRYNIDMGYLNTDVRSRIWLTYRKNFPKIGGTGPTTDSGWGCMLRC<br />
GQMMLAQALWQWDPENEYMQILEAFLDKKDSLYSIHQIAQMGVSE-GKAVGSWFGPNTVAQVLKKL----SA<br />
FDDWSSLCLHVAMDNTVIIEDISN---WRPLVLFIPLRLGLTEM-NVVYNEPLKACFTFKQSLGIIGGRPNH<br />
ATYFIGYFGNNLVYLDPHTTQQTVN-PDELSRIPDGSFHCVYPCRMNIADVDPSVALGFFCKSEEDFDDLCQ<br />
QIKKIIDGKSRPMFEIAK--------DRPQHWPVLE----</p>
<p>Clustal example</p>
<p><span class="code">CLUSTAL<br />
    <br />
QUERY/1-328           MEWLKIALVAFSGRGNDVDYPCSILGRVANNDKELVNILRNGFFLLTYRMNFSPLPHSSV<br />
UniRef90_Q4VBK1/1-294 ---------------PDTDEPVWILGACYNVKTKKSELLSDSRLWFTYRKKFSPIGGTGP<br />
UniRef90_A7RTH8/1-303 ---MDAACVTYEGVSHETEEDVWILGKRYNIDMGYLNTDVRSRIWLTYRKNFPKIGGTGP<br />
<br />
QUERY/1-328           TSDKGWGCLVRSSQMLLAHALWRYSANDCRLDHFRDMDTEDSTPFSLHKMVRAVMKKADV<br />
UniRef90_Q4VBK1/1-294 SSDAGWGCMLRCGQMILAQALWRWDPEKHQPKEYQRFLDKKDSCYSIHQMAQMGVGEGKS<br />
UniRef90_A7RTH8/1-303 TTDSGWGCMLRCGQMMLAQALWQWDPENEYMQILEAFLDKKDSLYSIHQIAQMGVSEGKA<br />
<br />
QUERY/1-328           FRPEYWTPSQGCEAIRCCVNNAVDRKLIPPIRVVVCSQGCLLAREICSNLEFGTVLILAP<br />
UniRef90_Q4VBK1/1-294 VG-EWYGPN----TVAQVLKKLALFDDWNSLSVYVSMDNTVVIEDIKKLCDWRPLLLVIP<br />
UniRef90_A7RTH8/1-303 VG-SWFGPN----TVAQVLKKLSAFDDWSSLCLHVAMDNTVIIEDIS---NWRPLVLFIP<br />
<br />
QUERY/1-328           MRCGASRRMTQMMFFSLEHLLHSSACIGVVGGVPQRSYYILGTSGQRLLYLDPHCMTQEA<br />
UniRef90_Q4VBK1/1-294 LRMGIN-SINPVYIQALKECFKMPQSCGVLGGKPNLAYYFIGFIDDELIYLDPH-TTQQA<br />
UniRef90_A7RTH8/1-303 LRLGLT-EMNVVYNEPLKACFTFKQSLGIIGGRPNHATYFIGYFGNNLVYLDPH-TTQQT<br />
<br />
QUERY/1-328           LVSSHAEKAGVVTVTASLVKSVRWDCVDTSCFLGFLVDSFAEWLELRTCLEELQRRGMEQ<br />
UniRef90_Q4VBK1/1-294 VDTESGSAVDDQSFHCQRPHRMKITSLDPSVALGFFCKSEEDFDSWCDLVQQ-------E<br />
UniRef90_A7RTH8/1-303 VNPDELSRIPDGSFHCVYPCRMNIADVDPSVALGFFCKSEEDFDDLCQQIK--------K<br />
<br />
QUERY/1-328           LLCVDDGVAVGLVDEEIAGWPSEEDVAE<br />
UniRef90_Q4VBK1/1-294 LLKKRNLRMFELVEKHPSHWPPF-----<br />
UniRef90_A7RTH8/1-303 IIDGKSRPMFEIAKDRPQHWPVLE----</span> </p>
<p>The gaps in the alignment can be represented by   
  *, -, space character, X and . (a dot). Other, custom gap characters can also be defined.</p>
<h3><a name="out" id="out"></a>The output </h3>
<p>AACon supports two modes of the output the first and the default mode is       
RESULT_NO_ALIGNMENT. Only the method names and the conservation score for each position in the alignment are presented. The second output format is RESULT_WITH_ALIGNMENT, where the input, in the form of FASTA formatted alignment, is included at the beginning of the file, followed by the same output as RESULT_NO_ALIGNMENT.  </p>
<p>RESULT_NO_ALIGNMENT example is below: <br />
    <span class="code">#KABAT 0.75 0.75 0.75 0.375 0.375 0.375 0.938 0.375 0.938 0.375 0.37 0.3<br />
      #JORES 0 0 0 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1<br />
      #SCHNEIDER 0.667 0.667 0.667 0.333 0.333 0.333 0.754 0.333 0.754 0.333 0<br />
      #SHENKIN 0.779 0.779 0.779 0.46 0.46 0.46 0.845 0.46 0.845 0.46 0.46 0.4<br />
      #GERSTEIN 0.667 0.667 0.667 0.333 0.333 0.333 0.754 0.333 0.754 0.33 0.3<br />
      #TAYLOR_GAPS 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0.632 1 0 1 0.526 0.842 0 0.5<br />
      #TAYLOR_NO_GAPS 0 1 0.526 0 0.895 0 0.947 0 0 0.789 0 0.842 0.947 0 0.94<br />
      #ZVELIBIL 0.1 0.3 0.4 0.1 0.2 0.1 0.3 0.1 0.3 0.2 0.2 0.1 0.3 0 0.2 0.2 <br />
      #KARLIN 0.338 0.338 0.338 0.433 0.31 0.264 0.558 0.288 0.558 0.338 0.431<br />
      #ARMON 0.617 0.617 0.617 0.208 0.103 0.063 0.617 0.129 0.617 0.177 0.204<br />
      #THOMPSON 0.719 0.727 0.708 0.451 0.394 0.384 0.471 0.374 0.49 0.428 0.4<br />
      #NOT_LANCET 0.011 0.005 0.054 0.172 0.14 0.108 0.194 0.145 0.194 0.1 0.2<br />
      #MIRNY 0.667 0.667 0.667 0.754 0.333 0.754 0.754 0.754 0.754 0.333 0.754<br />
      #WILLIAMSON 0.884 0.953 0.971 0.503 0.511 0.445 0.645 0.552 0.503 0.522 <br />
      #LANDGRAF 0.752 0.726 0.847 0.539 0.436 0.25 0.433 0.476 0.547 0.308 0.6<br />
      #SANDER 0 0 0 0.16 0.107 0.093 0.133 0.027 0.16 0.12 0.2 0.107 0.17 0.01<br />
      #VALDAR 0 0 0 0.187 0.115 0.115 0.256 0.046 0.256 0.164 0.233 0.092 0.25<br />
      #SMERFS 0.625 0.625 0.625 0.625 0.446 0.3 0.232 0.232 0.232 0.232 0.232 </span></p>
RESULT_WITH_ALIGNMENT output example is below: 
<p class="code">&gt;QUERY<br />
  MEWLKIALVAFSGRGNDVDYPCSILGRVANNDKELVNILRNGFFLLTYRMNFSPLPHSSVTSDKGWGCLVRSSQMLLAHA<br />
  LWRYSANDCRLDHFRDMDTEDSTPFSLHKMVRAVMKKADVFRPEYWTPSQGCEAIRCCVNNAVDRKLIPPIRVVVCSQGC<br />
  LLAREICSNLEFGTVLILAPMRCGASRRMTQMMFFSLEHLLHSSACIGVVGGVPQRSYYILGTSGQRLLYLDPHCMTQEA<br />
  LVSSHAEKAGVVTVTASLVKSVRWDCVDTSCFLGFLVDSFAEWLELRTCLEELQRRGMEQLLCVDDGVAVGLVDEEIAGW<br />
  PSEEDVAE<br />
  &gt;UniRef90_Q4VBK1<br />
  ---------------PDTDEPVWILGACYNVKTKKSELLSDSRLWFTYRKKFSPIGGTGPSSDAGWGCMLRCGQMILAQA<br />
  LWRWDPEKHQPKEYQRFLDKKDSCYSIHQMAQMGVGE-GKSVGEWYGPNTVAQVLKKL----ALFDDWNSLSVYVSMDNT<br />
  VVIEDIKKLCDWRPLLLVIPLRMGINS-INPVYIQALKECFKMPQSCGVLGGKPNLAYYFIGFIDDELIYLDPHTTQQA-<br />
  VDTESGSAVDDQSFHCQRPHRMKITSLDPSVALGFFCKSEEDFDSWCDLVQ-------QELLKKRNLRMFELVEKHPSHW<br />
  PPF-----<br />
  &gt;UniRef90_A7RTH8<br />
  ---MDAACVTYEGVSHETEEDVWILGKRYNIDMGYLNTDVRSRIWLTYRKNFPKIGGTGPTTDSGWGCMLRCGQMMLAQA<br />
  LWQWDPENEYMQILEAFLDKKDSLYSIHQIAQMGVSE-GKAVGSWFGPNTVAQVLKKL----SAFDDWSSLCLHVAMDNT<br />
  VIIEDISN---WRPLVLFIPLRLGLTEM-NVVYNEPLKACFTFKQSLGIIGGRPNHATYFIGYFGNNLVYLDPHTTQQTV<br />
  N-PDELSRIPDGSFHCVYPCRMNIADVDPSVALGFFCKSEEDFDDLCQQIKKIIDGKSRPMFEIAK--------DRPQHW<br />
  PVLE----<br />
  #KABAT 3 3 3 6 6 6 1.5 6 1.5 6 6 6 1.5 6 6 9 3 3 3 3 3 3 3 1 1 1 9 9 3 1 9 3 9 9 9 9 3 9 3 9 9 3 3 6 6 6 6 6 9 9 3 3 9 9 9 9 6 6 6 6 6 1.5 1.5 6 9 9 3 9 3 1 1 9 9 1.5 3 3 3 3 <br />
  #SANDER -72 -72 -72 -36 -48 -51 -42 -66 -36 -45 -27 -48 -33 -54 -45 3 30 9 30 -12 12 0 39 45 54 45 -18 -36 -72 -72 -72 -72 </p>
<p class="attention">Please note that only a small part of the output is presented in all examples that follow  for the sake of simplicity. Therefore the results are for illustration only. Normally each position of the alignment is given a conservation score by each  calculation method. </p>
<p><span class="hightlight">-f</span> option can be used to specify the output format. For example the command </p>
<p><span class="code">java -jar compbio-conservation-1.0.jar -i=data.align -m=KABAT -f=RESULT_WITH_ALIGNMENT
</span></p>
outputs the input alignment in to the result file. 
<h3><a name="run" id="run"></a>Calculating conservation</h3>
<p><strong>Calculate conservation using KABAT conservation method </strong></p>
<p>The alignment is read from the data.align file. </p>
<p class="code">java -jar compbio-conservation-1.0.jar -i=data.align -m=KABAT</p>
<p>The above command will print the results to the console. You should see something like this: </p>
<p><span class="code">#KABAT 3 3 3 6 6 6 1.5 6 1.5 6 6 6 1.5 6 6 9 3 3 3 3 3 3 3 1 1 1 9 9 </span></p>
<p><strong>Calculate conservation using two or more methods, for example SANDER and KABAT  </strong></p>
<p>Just specify the list of comma separated methods after -m switch like this: </p>
<p><span class="code">java -jar compbio-conservation-1.0.jar -i=data.align -m=SANDER,KABAT </span></p>
<p>The results will look something like this: </p>
<p><span class="code">#KABAT 3 3 3 6 6 6 1.5 6 1.5 6 6 6 1.5 6 6 9 3 3 3 3 3 3 3 1 1 1 9 9 3  <br />
  #SANDER -72 -72 -72 -36 -48 -51 -42 -66 -36 -45 -27 -48 -33 -54 -45 3 30 </span></p>
<p><strong>Calculate conservation using all supported conservation methods and make results comparable </strong></p>
<p><span class="code">java -jar compbio-conservation-1.0.jar -i=data.align -n </span>The results will be printed to the console. Where -n <a href="#normal">normalizes</a> all results.  </p>
<h3><a name="gap" id="gap"></a>Using custom gap character </h3>
<p>Assuming that the gaps in the alignment are represented by # and 1 symbols the following command will interpret them correctly. </p>
<p><span class="code">java -jar compbio-conservation-1.0.jar -i=data.align -m=KABAT -g=#,1</span></p>
<h3><a name="cpulim" id="cpulim"></a>Limiting AACon to X number of CPU only. </h3>
<p>By default AACon uses all CPUs available on the computer to perform calculations. This may be undesirable in the server environment. To restrict AACon to X CPU only use <span class="hightlight">-t</span> switch. </p>
<p class="code">java -jar compbio-conservation-1.0.jar -i=data.align -n -t=1 </p>
<p>The above command restricts AACon to a single CPU. </p>
<h3><a name="smerfs" id="smerfs"></a>Running SMERFS with custom parameters </h3>
<p>Unlike other methods, SMERFS supports a few custom parameters, in particular</p>
<ol>
  <li> The window width - an integer and an odd number</li>
  <li>Two methods of window scores to columns allocation:<br />
    MID_SCORE - gives the window score to the middle column<br />
    MAX_SCORE - gives the column the highest score of all the windows it <br />
    belongs to</li>
  <li>A gap percentage cutoff - a float greater than 0 and smaller or equal 1<br />
  </li>
  </ol>
<p>For example:  </p>
<p class="code"> java -jar compbio-conservation-1.0.jar -i=data.align -m=SMERFS -s=5,MID_SCORE,0.1</p>
<h3><a name="fileout" id="fileout"></a>Recording the output to the file </h3>
<p>Use <span class="hightlight">-o</span> option to print the results to the file instead of a console.  For example </p>
<p><span class="code">java -jar compbio-conservation-1.0.jar -i=data.align -n -o=outfile.txt</span></p>
<p>will produce output.txt results file. Nothing will be printed to the console. </p>
<h3><a name="execdet" id="execdet"></a>Execution details </h3>
<p>Use <span class="hightlight">-d</span> option to tell AACon to output its execution details.</p>
<p class="code">java -jar compbio-conservation-1.0.jar -i=data.align -n -d=stat.out -f=RESULT_WITH_ALIGNMENT</p>
<p> The output of the previous command, the context of the <span class="hightlight">stat.out</span> file is below</p>
<p class="code">No methods are request assuming all are required.<br />
  No output file is provided, writing results to the standard output.<br />
  Setting output format to RESULT_WITH_ALIGNMENT<br />
  Using 4 CPUs<br />
  Start time: 2010/12/08 12:02:35<br />
  Alignment loaded in: 115 ms<br />
  Alignment has: 3 sequences.<br />
  Alignment length is: 328 <br />
  KARLIN 15 ms<br />
  LANDGRAF 25 ms<br />
  SANDER 2 ms<br />
  VALDAR 2 ms<br />
  SMERFS 3 ms<br />
  SCHNEIDER 1 ms<br />
  KABAT 1 ms<br />
  SHENKIN 3 ms<br />
  JORES 5 ms<br />
  GERSTEIN 3 ms<br />
  ARMON 1 ms<br />
  THOMPSON 4 ms<br />
  NOT_LANCET 4 ms<br />
  ZVELIBIL 13 ms<br />
  MIRNY 4 ms<br />
  TAYLOR_GAPS 20 ms<br />
  WILLIAMSON 4 ms<br />
  TAYLOR_NO_GAPS 21 ms<br />
  Total calculation time: 0 s<br />
  End time: 2010/12/08 12:02:36</p>
<h3><a name="normal" id="normal"></a>Results normalization </h3>
Different conservation algorithms produce vastly different values. To compare them meaningfully one need to bring the results into the same range. AACon bring the results to the range between 0 and 1 if <span class="hightlight">-n</span> switch is used.
<h3><a name="largeal" id="largeal"></a>Conservation for large alignments (Java VM running out of memory) </h3>
<p>To enable AACon to deal with large alignments (thousands of sequences) let the Java Virtual Machine use more memory with the flag <span class="hightlight">-Xmx&lt;AMOUNT_OF_MEMORY&gt;</span> to your command line as follows: </p>
<p class="code">java -Xmx1G -jar aacon.jar &lt;options&gt; </p>
<p>Where 1G = 1 gigabyte of memory, the same result can be achieved with the following instruction<span class="hightlight"> -Xmx1000M</span> . However, floating point numbers like -Xmx1.5G are not allowed.</p>
<p>&nbsp; </p>
</div>
<!-- content end--> 
<div id="copyright">Last update: 10 December 2010<br/>
Peter Troshin and Geoff Barton, The Barton Group, University of Dundee, UK</div>
</div><!-- wrapper end-->
</div> <!-- page end-->
<!-- Google analitics -->
<script type="text/javascript">
var gaJsHost = (("https:" == document.location.protocol) ? "https://ssl." : "http://www.");
document.write(unescape("%3Cscript src='" + gaJsHost + "google-analytics.com/ga.js' type='text/javascript'%3E%3C/script%3E"));
</script>
<script type="text/javascript">
try{
var pageTracker = _gat._getTracker("UA-5356328-1");
pageTracker._trackPageview();
} catch(err) {}
</script>
</body>
</html>

