<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=iso-8859-1" />
<meta name="Last-modified" content="Mon, 11 Dec 2010 01:03:33 GMT"/>
<title>Amino Acid Conservation Service (AAConWS) Web Service</title>
<link href="ws.css" rel="stylesheet" type="text/css" media="screen,  projection, handheld, tv" />
<link rel="stylesheet" type="text/css" media="print" href="print.css"/>

<script type="text/javascript" src="prototype-1.6.0.3.js"></script>
</head>
<body>
<div id="page">
<div id="banner"><table> 
<tr><td style="width:130px;"><a href="http://www.dundee.ac.uk"><img class="logo" src="images/uod_lt.gif"  alt="University of Dundee"  title="University of Dundee" longdesc="http://www.dundee.ac.uk"/></a></td>
<td class="bg"><h2><span style="color:#1ACC1A" class="headeru">A</span>mino <span style="color:#80B3E6" class="headeru">A</span>cid <span style="color:#AE5439" class="headeru">Con</span>servation Service</h2></td>
</tr>
</table>
</div><!-- banner end-->
<div id="wrapper">
<div id="panel">
	<a href="index.html">Home</a>
<a href="command.html" title="Click to view more information on the command line version of AACon" >Standalone</a>
	<a class="selected" href="webservice.html" title="Click to view more information on AACon SOAP web service and how to use it" >Web Service</a>
	<a title="Click to view more information on AACon Java library" href="library.html">Library</a>	
	<a href="download.html" title="Click to download AACon" >Download</a>

	<a href="http://www.compbio.dundee.ac.uk">Barton Group</a></div>
<!-- panel end-->
<div id="content">

<h2 id="headtitle">AACon Web Service How To</h2>
<p>This page provides some tips on using AACon web service. Please note that for now all the examples are in Java. Other languages will follow given a sufficient demand. Please also note that AACon web service client requires Java 6. </p>
<strong>Outline</strong><ul><li><a href="#wsdl">AACon web service Details </a></li>
  <li><a href="#wsclient">Using web service command line client </a></li>
  <li><a href="#templclient">Structure of the template command line client</a></li>
  <li><a href="#wsfunctions">AACon web service functions overview</a></li>
  <li><a href="#webart">Building web services artifacts</a></li>
  <li><a href="#conn">Connecting to AAConWS</a></li>
  <li><a href="#runconserv">Calculating conservation </a></li>
  <li><a href="#checkstat">Checking the status of the calculation </a></li>
  <li><a href="#runconservpre">Calculating conservation with presets</a></li>
  <li><a href="#runconservpar">Calculating conservation with custom parameters</a></li>
  <li><a href="#complex">A complete AAConWS web service client example </a></li>
  </ul>
<hr />
<h4><a name="wsdl" id="wsdl"></a>AACon Web Service Details </h4>
<table>
  <tbody>
    <tr>
      <th width="24%">Feature</th>
      <th width="76%">Description</th>
    </tr>
    <tr>
      <td>Operated by </td>
      <td>The University of Dundee, Computational Biology Group headed by Prof Geoff Barton.  It is backed up by the College of
        Life Sciences HPC cluster. </td>
    </tr>
    <tr>
      <td><acronym title="Uniform Resource Locator">Service web site </acronym> </td>
      <td><a href="http://www.compbio.dundee.ac.uk/jabaws" title=
"http://www.compbio.dundee.ac.uk/jabaws" rel=
"nofollow">http://www.compbio.dundee.ac.uk/aacon</a></td>
    </tr>
    <tr>
      <td>Execution limits</td>
      <td>Tasks exceeding 5000&times;1000
        (sequences per letters) will not be accepted for alignment. If you
        would like to work with bigger alignments consider using a command line version of 
        AACon.</td>
    </tr>
    <tr>
      <td>AACon web service WSDL </td>
      <td><a href="http://www.compbio.dundee.ac.uk/aacon/AAConWS?wsdl"
title="http://www.compbio.dundee.ac.uk/jabaws/ClustalWS?" rel=
"nofollow">http://www.compbio.dundee.ac.uk/aacon/AAConWS?wsdl</a></td>
    </tr>
  </tbody>
</table>
<h4><a name="wsclient" id="wsclient"></a>Using Web Service Command Line Client</h4>
<p>The command client can be used to calculate conservation using public AACon web service.  The client is operating system independent and supports most of the AACon web service functions. Using this client you could calculate conservation  using web service features such as presets and parameters. Below is the list of options supported by the command line client. </p>
<span class="code"><strong>Usage:</strong> java -jar &lt;path_to_jar_file&gt; ACTION [OPTIONS]
<br />
<strong>ACTIONS:</strong><br />
-i=&lt;inputFile&gt; - full path to Fasta or Clustal formatted alignment file<br />
-parameters - lists parameters supported by web service<br />
-presets - lists presets supported by web service<br />
-limits - lists web services limits<br />
Please note that if input file is specified other actions are ignored
</p>
<br />
<strong> OPTIONS:</strong> (only for use with -i action):<br />
-r=&lt;presetName&gt; - name of the preset to use<br />
-o=&lt;outputFile&gt; - full path to the file where to write the results <br />
-f=&lt;parameterInputFile&gt; - the name of the file with the list of parameters to use.<br />
Please note that -r and -f options cannot be used together. Conservation can be calculated with either a preset or  the parameters from the file, but not both!</span>
<p>Using the command line web service client is easy. For example to calculate conservation for the alignment from <span class="hightlight">input.fasta</span> file using AACon web service with default settings and print the results to the console. By default, the conservation is calculated by SHENKIN method. </p>
<p class="code">java -jar aacon-client.jar -i=d:\input.fasta</p>
<p>Content of<span class="hightlight"> input.fasta</span> 
file is show below (please note sequences has been trimmed for clarity)
  <span class="code">&gt;QUERY  <br />
  MEWLKIALVAFSGRGNDVDYPCSILGRVANNDKELVNIL 
  <br />
  &gt;Q4VBK1 <br />
  ---------------PDTDEPVWILGACYNVKTKKSELL 
  <br />
  &gt;A7RTH8 <br />
  ---MDAACVTYEGVSHETEEDVWILGKRYNIDMGYLNTD </span>
</p>
Calculate conservation using 12 fast conservation methods (using preset &quot;Quick conservation&quot;), write output results to the result.txt file
<p><span class="code">java -jar aacon-client.jar -i=d:\input.fasta -o=d:\result.txt -r=&quot;Quick conservation&quot; </span></p>
<p>Get the list of presets supported by AACon web service  <br />
    <span class="code"> java -jar aacon-client.jar  -presets</span></p>
<h4><a name="templclient" id="templclient"></a>Structure of the template command line client</h4>
<table width="100%" border="1">
  <tr>
    <td width="18%"><strong>Packages</strong></td>
    <td width="82%"><strong>Classes and Interfaces </strong></td>
  </tr>
  <tr>
    <td>compbio.data.msa </td>
    <td>Annotation interface for all scoring methods  </td>
  </tr>
  <tr>
    <td>compbio.data.sequence</td>
    <td>AAConWS data types </td>
  </tr>
  <tr>
    <td>compbio.metadata</td>
    <td>AAConWS meta data types </td>
  </tr>
  <tr>
    <td>compbio.ws.client</td>
    <td>AAConWS command line client package with AAConClient.java class containing the main method  </td>
  </tr>
</table>
<p>  Please refer to a <a href="javadoc/index.html">data model javadoc</a> for a detailed description of each class and its methods. </p>
<h4><a name="wsfunctions" id="wsfunctions"></a>AACon web service functions overview </h4>
<p><strong>Functions for conservation calculation </strong><span class="code"> String id = analize(List&lt;FastaSequence&gt; list)<br />
  String id = customAnalize(List&lt;FastaSequence&gt; sequenceList, List&lt;Option&gt; optionList)<br />
  String id = presetAnalize(List&lt;FastaSequence&gt; sequenceList, Preset preset)</span></p>
<p><strong>Functions pertaining to job monitoring and control</strong><br />
    <span class="code">JobStatus status = getJobStatus(String id)<br />
      HashSet&lt;Score&gt; conservation = getConservation(String id)<br />
      boolean cancelled = cancelJob(String id)<br />
      ChunkHolder chunk = pullExecStatistics(String id, long marker)</span></p>
<p><strong>Functions relating to service features discovery</strong><br />
    <span class="code">RunnerConfig rc = getRunnerOptions()<br />
      Limit limit = getLimit(String name)<br />
      LimitsManager lm = getLimits()<br />
      PresetManager pm = getPresets()</span></p>
<p>Please refer to a <a href="javadoc/dm_javadoc/compbio/data/msa/Annotation.html">data model  javadoc</a> for a detailed description of each methods. </p>
<h4><a name="webart" id="webart"></a>Building web services artifacts</h4>
<p>AAConWS are the standard <a href="http://jax-ws.java.net/">JAX-WS</a> SOAP web services, which are <a href="http://www.ws-i.org/">WS-I</a> basic   profile compatible. This means that you could use your favourite programming language to work with AAConWS. Below is how you can generate portable artifacts to work with AAConWS from Java. However,  if programming in Java we recommend using our <a href="download.html"> client library</a> as it provides a handful of useful methods in addition to plain data types. </p>
<p class="code">wsimport -keep http://www.compbio.dundee.ac.uk/aacon/AAConWS?wsdl</p>
<h4><a name="conn" id="conn"></a>Connecting to AAConWS</h4>
<p>All the examples below assume that AACon web service command line client is in the classpath. You can download it from the <a href="download.html">download page</a>.  The code excerpt below will connect your program to AAConWS web service deployed in the University of Dundee. </p>
<p class="code"> Annotation&lt;AAConWS&gt; client = AAConWSClient.connect();</p>
<p>If you want to work with the <a href="#webart">generated artefacts</a> directly you can inspect AAConWS command line client source code and use it as a template for building your own custom AAConWS clients. </p>
<h4><a name="runconserv" id="runconserv"></a>Calculating conservation  </h4>
<p>Given that <span class="hightlight">client</span> is web service proxy, created as described in &quot;Connecting to AAConWS&quot; section, the conservation scores can be obtained as follows: </p>
<p class="code">1) List&lt;FastaSequence&gt; fsl = SequenceUtil.readFasta(new FileInputStream(&quot;alignment.fasta&quot;));<br />
  2) String jobId = client.analize(fsl);<br />
  3) HashSet&lt;Score&gt; result = client.getAnnotation(jobId);</p>
<p>Line  one loads sequence alignment from the file<br />
  Line two submits them to web service represented by msaws proxy <br />
  Line three retrieves the conservation scores caclulated according to Shenkin algorithm.  This line blocks the execution until the result is available. Use this with caution. In general, you should make sure that the calculation has been completed before attempting retrieving results. This is to avoid keeping the connection to the server on hold for a prolonged periods of time. While this may be ok with your local server, our public server (<a href="http://www.compbio.dundee.ac.uk/jabaws">www.compbio.dundee.ac.uk/aacon</a>) will not let you hold the connection for longer than 10 minutes. This is done to prevent excessive load on the server. The next section describes how to check the status of the calculation.<br />
  Methods and classes mentioned in the excerpt are available from the AAConWS client library. </p>
<h4><a name="checkstat" id="checkstat"></a>Checking the status of the calculation </h4>
<p> You may have noticed that there was no pause between submitting the job and retrieving the results. This is because <span class="hightlight">getResult(jobId)</span> method block the processing until the calculation is completed. However, taking into account that the connection holds server resources, our public server (<a href="http://www.compbio.dundee.ac.uk/jabaws">www.compbio.dundee.ac.uk/aacon</a>) is configured to reset the connection after 10 minutes of waiting. To work around the connection reset you are encouraged to check whether the calculation has been completed before accessing the results.	You can do it like this: </p>
<p> <span class="code">while (client.getJobStatus(jobId) != JobStatus.FINISHED) {<br />
  &nbsp;&nbsp;&nbsp;&nbsp;Thread.sleep(2000); // wait two  seconds, then recheck the status <br />
  }</span></p>
<h4><a name="runconservpre" id="runconservpre"></a>Calculating conservation with  presets</h4>
<p class="code">1) PresetManager&lt;AACon&gt; presets = client.getPresets();<br />
2) String jobId = client.presetAnalize(fsl, presets.getPresetByName(&quot;Quick conservation&quot;));<br />
3) HashSet&lt;Score&gt; result = client.getAnnotation(jobId);</p>
<p>Line one obtains the lists of presets supported by a web service.<br />
  Line two calls web service presetAnalise method with a chosen preset. This call returns a job identifier.<br />
  Lines three retrieves the results using job identifier.</p>
<p>Available presets are </p>
<ul>
  <li>&quot;Quick conservation&quot; (a collection of 12 fast conservation algorithms includes KABAT, JORES, SCHNEIDER, SHENKIN, GERSTEIN, TAYLOR_GAPS, TAYLOR_NO_GAPS, ZVELIBIL,ARMON, THOMPSON, NOT_LANCET, MIRNY, WILLIAMSON)</li>
  <li>&quot;Slow conservation&quot; (a collection of time consuming conservation algorithms includes LANDGRAF, KARLIN, SANDER, VALDAR and SMERFS) </li>
  <li>&quot;Complete conservation&quot; (all available algorithms) </li>
  </ul>
<h4><a name="runconservpar" id="runconservpar"></a>Calculating conservation with custom parameters</h4>
<p>Below is the example of using custom parameters for SMERFS method. </p>
<p class="code">// Using options<br />
1) RunnerConfig&lt;AACon&gt; options = client.getRunnerOptions();<br />
2) options.getArgument(&quot;Calculation method&quot;).setDefaultValue(&quot;SMERFS&quot;);<br />
3) options.getArgument(&quot;SMERFS Column Scoring Method&quot;).setDefaultValue(&quot;MAX_SCORE&quot;);<br />
4) options.getArgument(&quot;SMERFS Gap Threshhold&quot;).setDefaultValue(&quot;1&quot;);<br />
5) String jobId = client.customAnalize(fsl, options.getArguments());<br />
6) HashSet&lt;Score&gt; result = client.getAnnotation(jobId);</p>
<p>Line one obtains the RunnerConfig object that holds information on supported parameters and their values<br />
  Line two retrieve a particular parameter from the holder by its name and sets the new value for this parameter. <br />
  Lines three and four do the same but for two more parameters<br />
  Line five submit a job to a web service <br />
  Line six retrieves the results of the analisys. The names of all the parameters supported by a web service  can be obtained using options.getArguments() method. Further details on the methods available from RunnerConfig object are available from the <a href="javadoc/index.html">javadoc</a>. </p>
<h4><a name="complex" id="complex"></a>A complete AAConWS web service client example </h4>
<p>Finally, a complete example of the program that connects to AAConWS service is below. Three is also a <a href="AAConWSClientExample.pdf">PDF version</a> of this example with syntax highlighted. The text notes are commented by block style comments e.g. /* comment */, the code alternatives are commented out with the line  comments - &quot;//&quot;. You may want to remove line style comments to test alternatives of the functions. All you need for this to work is a <a href="download.html">AAConWS binary client</a>. Please make sure that the client is in the Java class path before running this example.</p>
<pre class="code" style="line-height:1em;">
import java.io.ByteArrayInputStream;
import java.io.FileNotFoundException;
import java.io.IOException;
import java.util.List;
import java.util.Set;

import compbio.data.msa.Annotation;
import compbio.data.sequence.FastaSequence;
import compbio.data.sequence.Score;
import compbio.data.sequence.SequenceUtil;
import compbio.metadata.JobSubmissionException;
import compbio.metadata.Preset;
import compbio.metadata.PresetManager;
import compbio.metadata.ResultNotAvailableException;
import compbio.metadata.UnsupportedRuntimeException;
import compbio.metadata.WrongParameterException;
import compbio.runner.conservation.AACon;


/**
 * AAConWS client example
 */
public class AAConWSClientExample {

	/*
	 * Input sequences for alignment. For the simplicity keep them in the class
	 */
	static final String input = &quot;&gt;Foo      \r\n&quot;
			+ &quot;MTADGPRELLQLRAAVRHRPQDFVAWLMLADAELGMGDTTAGEMAVQRGLALHPGHPEAV\r\n&quot;
			+ &quot;ARLGRVRWTQQRHAEAAVLLQQASDAAPEHPGIALWLGHALEDAGQAEAAAAAYTRAHQL\r\n&quot;
			+ &quot;LPEEPYITAQLLNWRRRLCDWRALDVLSAQVRAAVAQGVGAVEPFAFLSEDASAAEQLAC\r\n&quot;
			+ &quot;ARTRAQAIAASVRPLAPTRVRSKGPLRVGFVSNGFGAHPTGLLTVALFEALQRRQPDLQM\r\n&quot;
			+ &quot;HLFATSGDDGSTLRTRLAQASTLHDVTALGHLATAKHIRHHGIDLLFDLRGWGGGGRPEV\r\n&quot;
			+ &quot;FALRPAPVQVNWLAYPGTSGAPWMDYVLGDAFALPPALEPFYSEHVLRLQGAFQPSDTSR\r\n&quot;
			+ &quot;VVAEPPSRTQCGLPEQGVVLCCFNNSYKLNPQSMARMLAVLREVPDSVLWLLSGPGEADA\r\n&quot;
			+ &quot;RLRAFAHAQGVDAQRLVFMPKLPHPQYLARYRHADLFLDTHPYNAHTTASDALWTGCPVL\r\n&quot;
			+ &quot;TTPGETFAARVAGSLNHHLGLDEMNVADDAAFVAKAVALASDPAALTALHARVDVLRRES\r\n&quot;
			+ &quot;GVFEMDGFADDFGALLQALARRHGWLGI\r\n&quot;
			+ &quot;\r\n&quot;
			+ &quot;&gt;Bar                    \r\n&quot;
			+ &quot;-----------------------------------MGDTTAGEMAVQRGLALH-------\r\n&quot;
			+ &quot;---------QQRHAEAAVLLQQASDAAPEHPGIALWL-HALEDAGQAEAAAA-YTRAHQL\r\n&quot;
			+ &quot;LPEEPYITAQLLN--------------------AVAQGVGAVEPFAFLSEDASAAE----\r\n&quot;
			+ &quot;----------SVRPLAPTRVRSKGPLRVGFVSNGFGAHPTGLLTVALFEALQRRQPDLQM\r\n&quot;
			+ &quot;HLFATSGDDGSTLRTRLAQASTLHDVTALGHLATAKHIRHHGIDLLFDLRGWGGGGRPEV\r\n&quot;
			+ &quot;FALRPAPVQVNWLAYPGTSGAPWMDYVLGDAFALPPALEPFYSEHVLRLQGAFQPSDTSR\r\n&quot;
			+ &quot;VVAEPPSRTQCGLPEQGVVLCCFNNSYKLNPQSMARMLAVLREVPDSVLWLLSGPGEADA\r\n&quot;
			+ &quot;RLRAFAHAQGVDAQRLVFMPKLPHPQYLARYRHADLFLDTHPYNAHTTASDALWTGCPVL\r\n&quot;
			+ &quot;TTPGETFAARVAGSLNHHLGLDEMNVADDAAFVAKAVALASDPAALTALHARVDVLRRES\r\n&quot;
			+ &quot;GVFEMDGFADDFGALLQALARRHGWLGI\r\n&quot;
			+ &quot;\r\n&quot;
			+ &quot;&gt;Noname             \r\n&quot;
			+ &quot;-MTADGPRELLQLRAAVRHRPQDVAWLMLADAELGMGDTTAGEMAVQRGLALHPGHPEAV\r\n&quot;
			+ &quot;ARLGRVRWTQQRHAEAAVLLQQASDAAPEHPGIALWLGHALED--------------HQL\r\n&quot;
			+ &quot;LPEEPYITAQLDVLSAQVR-------------AAVAQGVGAVEPFAFLSEDASAAEQLAC\r\n&quot;
			+ &quot;ARTRAQAIAASVRPLAPTRVRSKGPLRVGFVSNGFGAHPTGLLTVALFEALQRRQPDLQM\r\n&quot;
			+ &quot;HLFATSGDDGSTLRTRLAQASTLHDVTALGHLATAKHIRHHGIDLLFDLRGWGGGGRPEV\r\n&quot;
			+ &quot;FALRPAPVQVNWLAYPGTSGAPWMDYVLGDAFALPPALEPFYSEHVLRLQGAFQPSDTSR\r\n&quot;
			+ &quot;VVAEPPSRTQCGLPEQGVVLCCFNNSYKLNPQSMARMLAVLREVPDSVLWLLSGPGEADA\r\n&quot;
			+ &quot;RLRAFAHAQGVDAQRLVFMPKLPHPQYLARYRHADLFLDTHPYNAHTTASDALWTGCPVL\r\n&quot;
			+ &quot;TTPGETFAARVAGSLNHHLGLDEMNVADDAAFVAKAVALASDPAALTALHARVDVLRRES\r\n&quot;
			+ &quot;I---------------------------&quot;;

	public static void main(String[] args) throws UnsupportedRuntimeException,
			JobSubmissionException, WrongParameterException,
			FileNotFoundException, IOException, ResultNotAvailableException,
			InterruptedException {

		/*
		 * Annotation interface for AAConWS web service instance
		 */
		Annotation&lt;AACon&gt; client = AAConClient.connect();

		/* Get the list of available presets */
		PresetManager presetman = client.getPresets();

		/* Get the Preset object by preset name */
		Preset preset = presetman.getPresetByName(&quot;Complete conservation&quot;);

		/*
		 * Load sequences in FASTA format from the file You can use something
		 * like new FileInputStream() to load sequence from the file
		 */
		List&lt;FastaSequence&gt; fastalist = SequenceUtil
				.readFasta(new ByteArrayInputStream(input.getBytes()));

		/*
		 * Submit loaded sequences for an alignment using preset. The job
		 * identifier is returned by this method, you can retrieve the results
		 * with it sometime later.
		 */
		String jobId = client.presetAnalize(fastalist, preset);

		/* This method will block for the duration of the calculation */
		Set&lt;Score&gt; result = client.getAnnotation(jobId);

		/*
		 * This is a better way of obtaining results, it does not involve
		 * holding the connection open for the duration of the calculation,
		 * Besides, as the University of Dundee public server will reset the
		 * connection after 10 minutes of idling, this is the only way to obtain
		 * the results of long running task from our public server.
		 */
		// while (client.getJobStatus(jobId) != JobStatus.FINISHED) {
		// Thread.sleep(1000); // wait a second, then recheck the status
		// }

		/* Output the alignment to standard out */
		Score.write(result, System.out);

		/* Alternatively, you can record retrieved alignment into the file */
		// FileOutputStream out = new FileOutputStream(&quot;result.txt&quot;);
		// Score.write(result, out);
		// out.close();
	}
}

</pre>
For a more detailed description of all available types and their functions please refer to the <a href="javadoc/index.html">data model javadoc</a>.</div>
<!-- content end--> 
<div id="copyright">Last update: 10 December 2010<br/>
Peter Troshin and Geoff Barton, The Barton Group, University of Dundee, UK</div>
</div><!-- wrapper end-->
</div> <!-- page end-->
<!-- Google analitics -->
<!-- Google analitics -->
<script type="text/javascript">
var gaJsHost = (("https:" == document.location.protocol) ? "https://ssl." : "http://www.");
document.write(unescape("%3Cscript src='" + gaJsHost + "google-analytics.com/ga.js' type='text/javascript'%3E%3C/script%3E"));
</script>
<script type="text/javascript">
try{
var pageTracker = _gat._getTracker("UA-5356328-1");
pageTracker._trackPageview();
} catch(err) {}
</script>
</body>
</html>

